libs/depgraph/README.md

```diff
@@ -0,0 +1,296 @@
+
+Additional Benefits
+-------------------
+
+Dependency graph not only allows to solve a common problem in one place, therefore
+shortening the code size and the complexity of management agents, but it also enforces
+a much more readable, sustainable and extensible programming style. Consider the following
+(networking) example comparing a usual, overly imperative programming style and how it changes
+when depgraph is introduced:
+
+* Usual imperative programming:
+
+```text
+whenNetworkConfigChanges(<high-level network config>) {
+    determineObsoleteVlans()
+    removeObsoleteVlans()
+    determineObsoleteBonds()
+    removeObsoleteBonds()
+    changeInterfaceIPsIfNeeded()
+    ifSomethingChangedRestartDnsmasq()
+    addNewBonds()
+    addNewVlans()
+    ...
+}
+```
+
+* Declarative programming with depgraph:
+
+```text
+whenNetworkConfigChanges(<high-level network config>) {
+    newConfig := []ConfigItem{
+        interface(params),
+        arpEntry(params),
+        arpEntry(params),
+        route(params),
+        route(params),
+        bridge(params),
+        dnsmasq(params),
+        // A comment explaining why this config item is here…
+        iptablesChain(params),
+        iptablesChain(params),
+        ...
+    }
+    graph.Cluster(<network-name>).Put(newConfig)
+    err := graph.Sync()
+    ...
+}
+```
+
+Note that the new approach is not only easier for the developer and therefore less
+bug-prone, but also allows to explicitly express the intent (= `newConfig`), while
+the steps (the sequence of configuration changes) needed to take to get there are
+implicit. Compare that with the "usual" approach, where the steps are explicit,
+but the programmer's intent is implicit. To determine what the program is trying
+to configure, one must study the code thoroughly and build a mental image of the
+intended state. If the programmer made a mistake in that complex code, one might
+get a wrong idea of what the intended state is.
+
+Lastly, with the dependency graph, it is much easier to add new features.
+For the most part, a programmer only needs to implement handlers for new configuration
+items and describe their dependencies. The rest is being taken care of by the graph.
+
+Limitations and Future Plans
+----------------------------
+
+1. Currently, the synchronization procedure executes Create/Modify/Delete operations
+   sequentially and from a single Go routine. This may be a limiting factor if changing
+   the state of an item requires to perform a long-running task, such as downloading
+   a large file from the Internet.
+   A support for *asynchronous operation execution* is already planned for the near future,
+   however. Configurator will have an option (probably by use of the context argument)
+   to continue running operation inside a separate Go routine, exiting back to depgraph
+   immediately and informing that the state change continues in the background.
```

Comment:
FWIW this doesn't fit with how nim currently works where it asks domainmgr to get an interface back, when there is a change to AssignableAdapters in the nim event loop it will re-evaulate the current state and proceed.

It could be that we should rewrite and simplify nim to have a goroutine which does the application of a new DevicePortConfig, with the AssignableAdapter handler kicking such a goroutine. Is that what you have in mind?

Reply:
Actually for this kind of case I have prepared a different solution - items in the graph marked as "external", representing objects managed by other agents (microservices): https://github.com/milan-zededa/eve/blob/dependency-graph/libs/depgraph/depgraph_api.go#L169-L175
In the NIM depgraph there will two items in the graph for a device port as a whole, say "physio" and "system-adapter", where "physio" will be an external item, added/removed to/from the graph based on AssignableAdapters, and "system-adapter" will depend on "physio" and contain the interface configuration to apply. So Create of "system-adapter" doesn't actually "create" the interface, rather applies the desired interface configuration (like MTU etc.).

Reply:
Asynchronicity will be used for Create/Modify/Delete operations if they could take too long to complete (e.g. download a blob, untar an archive, copy something big, etc.) and would therefore delay other managed objects that do not even depend on it.

Reply:
I guess I don't see exactly what that flow would look like hence hard to comment.
As you "execute" the steps and get to an external, then how and who tracks that some event (from the asynch being done) will need to restart the execution at that point and presumably also capture the case when the external asynch fails or times out?

It seems like this executing can be complex. Currently how tied is the execution of the steps separable from the determination of the steps needed to reconcile? If it is possible to just ask the reconciler for the steps and separately execute them (including tracking what remains to be executed), then it might be easier to handle the asynch steps.

But perhaps you've already figured this out. Should we have a call to discuss?